src/main/java/com/amplitude/HttpTransport.java

```diff
@@ -155,36 +170,30 @@ private void retryEventsOnLoop(String userId, String deviceId) {
                       }
                     }
                     triggerEventCallbacks(eventsToDrop, retryResult.statusCode, "Invalid events.");
-                    eventCount -= numEventsRemoved;
-                    eventsInRetry.addAndGet(-numEventsRemoved);
-                  }
-                  if (!shouldRetry || eventCount < 1) {
-                    break;
-                  }
-                  if (shouldReduceEventCount && !isLastTry) {
-                    triggerEventCallbacks(
-                        eventsBuffer.subList(eventCount / 2, eventCount),
-                        retryResult.statusCode,
-                        "Event dropped for retry");
-                    eventCount /= 2;
-                  }
-                  if (isLastTry) {
-                    triggerEventCallbacks(
-                        eventsBuffer.subList(0, eventCount),
-                        retryResult.statusCode,
-                        "Event retries exhausted.");
-                  }
-                  if (eventCount == 0) {
-                    break;
+                  } else if (shouldReduceEventCount) {
+                    List<Event> eventsToDrop = eventsBuffer.subList(eventCount / 2, eventCount);
+                    triggerEventCallbacks(eventsToDrop, retryResult.statusCode, "Event dropped for retry");
+                    eventsBuffer = eventsBuffer.subList(0, eventCount / 2);
                   }
+
                 } catch (InterruptedException | AmplitudeInvalidAPIKeyException e) {
                   // The retry logic should only be executed after the API key checking passed.
                   // This catch AmplitudeInvalidAPIKeyException is just for handling
                   // retryEventsOnce in thread.
+                  logger.debug("RETRY", "Retry thread got interrupted");
                   Thread.currentThread().interrupt();
                 }
               }
-              eventsInRetry.addAndGet(-eventCount);
+              if (recordThrottledId) {
+                synchronized (throttleLock) {
+                  if (throttledUserId.containsKey(userId)) {
```

Comment:
The throttling logic here is in
getEventListToRetry(initial failure) -> add userId, deviceId -> wait 100ms -> start retry -> remove userId, deviceId from throttle after retry success or exhaust retries
Is that right?